extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,618 @@
+          iframeBox.width,
+      'allowedExpansion_b': viewportSize.height -
+          iframeBox.height,
+      'allowedExpansion_t': 0,
+      'allowedExpansion_l': 0,
+      'yInView': this.getPercInView(viewportSize.height,
+          iframeBox.top, iframeBox.bottom),
+      'xInView': this.getPercInView(viewportSize.width,
+          iframeBox.left, iframeBox.right),
+    });
+    this.currentGeometry_ = currentGeometry;
+    return JSON.stringify(currentGeometry);
+  }
+
+  /**
+   * Helper function to calculate both the xInView and yInView of the
+   * geometry update messages. In the case of a 400px wide viewport,
+   * with a 100px wide creative that starts at x position 50, if we
+   * are calculating xInView, rootBoundEnd is 400, boundingRectStart
+   * is 50, and boundingRectEnd is 150.
+   * @param {number} rootBoundEnd
+   * @param {number} boundingRectStart
+   * @param {number} boundingRectEnd
+   * @return {number}
+   */
+  getPercInView(rootBoundEnd, boundingRectStart, boundingRectEnd) {
+    const lengthInView = (boundingRectEnd >= rootBoundEnd) ?
+      rootBoundEnd - boundingRectStart : boundingRectEnd;
+    const percInView = lengthInView / (boundingRectEnd - boundingRectStart);
+    return Math.max(0, Math.min(1, percInView)) || 0;
+  }
+
+  /**
+   * Handles serializing and sending messages to the safeframe.
+   * @param {!Object} payload
+   * @param {string} serviceName
+   * @private
+   */
+  sendMessage_(payload, serviceName) {
+    dev().assert(this.iframe_.contentWindow,
+        'Frame contentWindow unavailable.');
+    const message = dict();
+    message[MESSAGE_FIELDS.CHANNEL] = this.channel;
+    message[MESSAGE_FIELDS.PAYLOAD] = JSON.stringify(
+        /** @type {!JsonObject} */(payload));
+    message[MESSAGE_FIELDS.SERVICE] = serviceName;
+    message[MESSAGE_FIELDS.SENTINEL] = this.sentinel_;
+    message[MESSAGE_FIELDS.ENDPOINT_IDENTITY] = this.endpointIdentity_;
+    this.iframe_.contentWindow./*OK*/postMessage(
+        JSON.stringify(message), SAFEFRAME_ORIGIN);
+  }
+
+  /**
+   * Routes messages to their appropriate handler.
+   * @param {!JsonObject} payload
+   * @param {string} service
+   */
+  processMessage(payload, service) {
+    // We are not logging unexpected messages, and some expected
+    // messages are being dropped, like init_done, as we don't need them.
+    switch (service) {
+      case SERVICE.CREATIVE_GEOMETRY_UPDATE:
+        this.handleFluidMessage_(payload);
+        break;
+      case SERVICE.EXPAND_REQUEST:
+        this.handleExpandRequest_(payload);
+        break;
+      case SERVICE.REGISTER_DONE:
+        this.isRegistered_ = true;
+        break;
+      case SERVICE.COLLAPSE_REQUEST:
+        this.handleCollapseRequest_();
+        break;
+      default:
+        break;
+    }
+  }
+
+  /**
+   * @param {!JsonObject} payload
+   * @private
+   */
+  handleExpandRequest_(payload) {
+    if (!this.isCollapsed_ || !this.isRegistered_) {
```

Comment:
in what cases would it not be registered?

Reply:
Not sure, but a previous comment from warren said that it needs to be checked